### ISSUE TITLE
Error when executing an asset via Databricks step launcher on Kubernetes

### ISSUE THREAD

**Issue by ske94** [dagster-io/dagster#20083]
### Dagster version

1.6.0

### What's the issue?

Our general setup is that we're using Dagster with Databricks. Until now, we've used our local development clients to setup and test all the Dagster code including the usage of the Databricks Step Launcher. Everything is working fine - we start jobs in Dagster containing multiple assets (e.g. one is doing stuff directly in Dagster and one is executed on Databricks).

Now, after starting to move our code into a Kubernetes cluster as a user deployment, we're experiencing an issue with the step launcher, as we want to run some assets using the Databricks step launcher. Generally everything is working fine on the cluster and we can execute other assets without issues. But when running with the step launcher, the following error occurs on Databricks:

![image](https://github.com/dagster-io/dagster/assets/101324379/4697c4b2-2906-49b4-b17b-eccb0dae351c)

The whole error has a pretty long stacktrace starting at the databricks_step_main.py, which is moved as part of the step launcher payload to databricks and is executed first with some arguments:
![image](https://github.com/dagster-io/dagster/assets/101324379/c28e5ac5-f342-4c4e-a6a7-4af294c4bc45)

In line 136 after setting up the external Dagster instance the function _run_step_from_ref_ is executed and ultimately results in an error. Our understandig is that the actual execution will take place at this point, but we can't really figure out why the job name is not suffixed with the expected "_0". On Kubernetes the Job is prefixed with "_0"

![image](https://github.com/dagster-io/dagster/assets/101324379/2562d8ca-e294-4dfa-8c05-3898b2dbc088)

The trace until the actual error goes across several parts of the Dagster code:

![image](https://github.com/dagster-io/dagster/assets/101324379/f5f760d3-82a4-4ad7-9e08-9475bbb4f53b)
![image](https://github.com/dagster-io/dagster/assets/101324379/70d1b0f5-1536-4178-afc6-ca2ca10d5ef1)
![image](https://github.com/dagster-io/dagster/assets/101324379/74366fed-77b8-4f28-a2f6-a3e2e097c35a)
![image](https://github.com/dagster-io/dagster/assets/101324379/ac3dd714-b8a6-4a32-8280-f7fb87586ce2)
![image](https://github.com/dagster-io/dagster/assets/101324379/696bcaaf-d2dd-479e-8e32-45426d677d0b)
![image](https://github.com/dagster-io/dagster/assets/101324379/2416dc27-3732-48f9-af11-fe07c79011ba)
![image](https://github.com/dagster-io/dagster/assets/101324379/cf07f770-ced4-42b6-a973-e9ca1abf5c94)

Something, we're also trying to understand is why it does work locally. So we made sure that the code that is zipped up and sent to Databricks is pretty much the same locally and in Kubernetes.

It would be great, if someone with deeper knowledge of the step launcher / external execution architecture could support in resolving this issue.

### What did you expect to happen?

For us this behaviour does not seem to be normal as we don't understand which difference between a local and a Kubernetes deployment makes the Databricks step launcher break at this point.

### How to reproduce?

The setup to reproduce is a Kubernetes Cluster running on AKS that is executing an asset using the Databricks step launcher on Azure Databricks.

### Deployment type

Dagster Helm chart

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by OwenKephart** [comment]
Hi @ske94 -- this is definitely odd behavior. Does the job you're launching contain any partitioned assets?

When you have a set of assets in a repository, behind the scenes, Dagster will create individual asset jobs, grouped by partitions definition. If you have a single partitions definition across all assets (or they're all unpartitioned), this will result in a single job named `__ASSET_JOB`. If you have more groups, you'll get jobs like `__ASSET_JOB_0`, `__ASSET_JOB_1`, etc.: https://sourcegraph.com/github.com/dagster-io/dagster@7deb4094fab3cdc1156b5c9c56d399c4c77b3d1f/-/blob/python_modules/dagster/dagster/_core/definitions/assets_job.py?L64:5-64:24

So it seems like, in the launching process' view of your code, there are multiple distinct definitions (leading to the suffix), but in Databricks' view of the world (i.e. the code that is zipped up and sent to it), there is only a single definition.

Another quick thing to check is that the library version of Dagster that's on your databricks cluster matches up to what you expect, but I think that's unlikely to make a difference in this specific case.

**Comment by ske94** [comment]
Hi @OwenKephart! Thanks for your reply!

Yes, the job contains partitioned assets (right now it is only one partition). Thats how it looks:

![image](https://github.com/dagster-io/dagster/assets/101324379/46c2e2ea-488c-447e-8116-b1d508d72f7b)

In the meantime we tried different stuff regarding the job. We realized that it doesn't make any difference if we execute the job (this leads to the job named like it is in our Dagster instance on Kubernetes - in our example kubernetes_test_job) or if we execute the two Databricks assets as a selection (this leads to the job named __ASSET_JOB_0 on Kubernetes). In both cases, this leads to the error above. That means that it is somehow always an issue, no matter how the job is named on the Kubernetes side.

We also tried to replace the job_name properties from the StepRunRef object that is created in the databricks_step_main.py, but this leads to another problem (which makes sense because it doesn't seem right to do this):
![image](https://github.com/dagster-io/dagster/assets/101324379/b80abd4a-f560-4b5f-a4e5-06db4510f02e)

After that, we also tried to remove the asset selection properties, but from this point I think this is going into the wrong direction:
![image](https://github.com/dagster-io/dagster/assets/101324379/c23f09ff-8e94-4b5d-affa-1f76766d797e)

**Comment by OwenKephart** [comment]
@ske94 The other piece of the puzzle here is the "Found '__ASSET_JOB'" part of the original error message, which seems to indicate that the _only_ job found in the Definitions that's constructed in Databricks. So maybe the issue here is actually that the code packaged up and sent to the Databricks cluster is somehow missing the relevant assets.

This plus the difference in behavior between local and k8s leads me to think that this could possible be an issue with how the relevant Docker image is being created / how the relevant files are being copied around? What do you have set as your `local_dagster_job_package_path`?

**Comment by ske94** [comment]
Hi @OwenKephart!

First of all, we found a solution! Your comment pointed us into the right direction. You were right that the external Dagster instance on Databricks missed the assets. We build assets based on YAML configuration files and messed with the zipping process compared to the local execution. So on Databricks the relevant assets were not present and the external step run probably couldn't map them to the assets from Kubernetes.

The creation of the __ASSET_JOB on Kubernetes probably came from a test asset which we did not build from config files. I guess without this test asset another more meaningful error would have occurred. 

Anyways, thank you very much for your help! :)
